File: seshat/tooling/typescript.py
"""
TypeScript/JavaScript language strategy for the tooling system.

Supports projects with package.json files, detecting and running
tools like ESLint, Biome, Prettier, TypeScript, Jest, and Vitest.
"""

import json
from pathlib import Path

from .base import (
    BaseLanguageStrategy,
    ToolCommand,
    ToolingConfig,
    SeshatConfig,
)


# File extensions for TypeScript/JavaScript projects
TS_JS_EXTENSIONS = {
    ".js", ".mjs", ".cjs", ".jsx",  # JavaScript
    ".ts", ".tsx", ".mts", ".cts",  # TypeScript (excluding .d.ts for lint)
}

# Extensions that should be type-checked (includes declaration files)
TYPECHECK_EXTENSIONS = {
    ".ts", ".tsx", ".mts", ".cts", ".d.ts", ".d.mts", ".d.cts",
}

# Patterns for test files
TEST_FILE_PATTERNS = {
    ".test.ts", ".test.js", ".test.tsx", ".test.jsx",
    ".spec.ts", ".spec.js", ".spec.tsx", ".spec.jsx",
}


class TypeScriptStrategy(BaseLanguageStrategy):
    """
    Strategy for TypeScript/JavaScript projects.
    
    Detects projects with package.json and discovers tools like
    ESLint, Biome, Prettier, TypeScript, Jest, and Vitest.
    """
    
    @property
    def name(self) -> str:
        return "typescript"
    
    @property
    def detection_files(self) -> list[str]:
        return ["package.json"]
    
    @property
    def lint_extensions(self) -> set[str]:
        return TS_JS_EXTENSIONS
    
    @property
    def typecheck_extensions(self) -> set[str]:
        return TYPECHECK_EXTENSIONS
    
    @property
    def test_patterns(self) -> set[str]:
        return TEST_FILE_PATTERNS
    
    @property
    def default_tools(self) -> dict[str, ToolCommand]:
        return {
            "eslint": ToolCommand(
                name="eslint",
                command=["npx", "eslint"],
                check_type="lint",
                pass_files=True,
            ),
            "biome": ToolCommand(
                name="biome",
                command=["npx", "@biomejs/biome", "check"],
                check_type="lint",
                pass_files=True,
            ),
            "prettier": ToolCommand(
                name="prettier",
                command=["npx", "prettier", "--check"],
                check_type="lint",
                pass_files=True,
            ),
            "tsc": ToolCommand(
                name="tsc",
                command=["npx", "tsc", "--noEmit"],
                check_type="typecheck",
                pass_files=False,  # tsc should check entire project
            ),
            "jest": ToolCommand(
                name="jest",
                command=["npx", "jest", "--passWithNoTests"],
                check_type="test",
                pass_files=True,
            ),
            "vitest": ToolCommand(
                name="vitest",
                command=["npx", "vitest", "run"],
                check_type="test",
                pass_files=False,
            ),
        }
    
    def discover_tools(
        self,
        path: Path,
        seshat_config: SeshatConfig,
    ) -> ToolingConfig:
        """Discover TypeScript/JavaScript tools from package.json."""
        config = ToolingConfig(project_type="typescript")
        package_json_path = path / "package.json"
        
        if not package_json_path.exists():
            return config
        
        try:
            with open(package_json_path, "r", encoding="utf-8") as f:
                pkg = json.load(f)
        except Exception:
            return config
        
        deps = {}
        deps.update(pkg.get("dependencies", {}))
        deps.update(pkg.get("devDependencies", {}))
        scripts = pkg.get("scripts", {})
        
        # Check for linters
        if "eslint" in deps or "@eslint/js" in deps:
            tool = self._get_tool_config("eslint", "lint", seshat_config)
            # Force pass_files for TypeScript lint
            tool.pass_files = True
            config.tools["lint"] = tool
        elif "@biomejs/biome" in deps:
            tool = self._get_tool_config("biome", "lint", seshat_config)
            tool.pass_files = True
            config.tools["lint"] = tool
        
        # Check for TypeScript
        if "typescript" in deps:
            tool = self._get_tool_config("tsc", "typecheck", seshat_config)
            if "typecheck" in scripts:
                tool.command = ["npm", "run", "typecheck"]
                tool.pass_files = False
            elif "type-check" in scripts:
                tool.command = ["npm", "run", "type-check"]
                tool.pass_files = False
            self._apply_command_overrides(
                tool, seshat_config.checks.get("typecheck", {}), seshat_config
            )
            config.tools["typecheck"] = tool
        
        # Check for test runners
        if "jest" in deps:
            tool = self._get_tool_config("jest", "test", seshat_config)
            if "test" in scripts:
                tool.command = ["npm", "run", "test"]
                tool.pass_files = False
            self._apply_command_overrides(
                tool, seshat_config.checks.get("test", {}), seshat_config
            )
            config.tools["test"] = tool
        elif "vitest" in deps:
            tool = self._get_tool_config("vitest", "test", seshat_config)
            if "test" in scripts:
                tool.command = ["npm", "run", "test"]
                tool.pass_files = False
            self._apply_command_overrides(
                tool, seshat_config.checks.get("test", {}), seshat_config
            )
            config.tools["test"] = tool
        
        return config

File: src/tooling/typescript.rs
use super::runner::{
    apply_overrides, collect_package_deps, deps_has, tool_from_default, LanguageStrategy,
};
use super::types::{ToolCommand, ToolingConfig};
use crate::config::ProjectConfig;
use serde_json::Value;
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

#[derive(Debug, Clone)]
pub(super) struct TypeScriptStrategy {
    root: PathBuf,
}

impl TypeScriptStrategy {
    pub(super) fn new(root: impl Into<PathBuf>) -> Self {
        Self { root: root.into() }
    }
}

impl LanguageStrategy for TypeScriptStrategy {
    fn name(&self) -> &'static str {
        "typescript"
    }

    fn detection_files(&self) -> &'static [&'static str] {
        &["package.json"]
    }

    fn lint_extensions(&self) -> &'static [&'static str] {
        &[".js", ".mjs", ".cjs", ".jsx", ".ts", ".tsx", ".mts", ".cts"]
    }

    fn typecheck_extensions(&self) -> &'static [&'static str] {
        &[".ts", ".tsx", ".mts", ".cts", ".d.ts", ".d.mts", ".d.cts"]
    }

    fn test_patterns(&self) -> &'static [&'static str] {
        &[
            ".test.ts",
            ".test.js",
            ".test.tsx",
            ".test.jsx",
            ".spec.ts",
            ".spec.js",
            ".spec.tsx",
            ".spec.jsx",
        ]
    }

    fn default_tools(&self) -> BTreeMap<&'static str, ToolCommand> {
        BTreeMap::from([
            (
                "eslint",
                ToolCommand::new("eslint", &["npx", "eslint"], "lint")
                    .with_files()
                    .with_fix(&["npx", "eslint", "--fix"]),
            ),
            (
                "biome",
                ToolCommand::new("biome", &["npx", "@biomejs/biome", "check"], "lint")
                    .with_files()
                    .with_fix(&["npx", "@biomejs/biome", "check", "--write"]),
            ),
            (
                "prettier",
                ToolCommand::new("prettier", &["npx", "prettier", "--check"], "lint")
                    .with_files()
                    .with_fix(&["npx", "prettier", "--write"]),
            ),
            (
                "tsc",
                ToolCommand::new("tsc", &["npx", "tsc", "--noEmit"], "typecheck"),
            ),
            (
                "jest",
                ToolCommand::new("jest", &["npx", "jest", "--passWithNoTests"], "test")
                    .with_files(),
            ),
            (
                "vitest",
                ToolCommand::new("vitest", &["npx", "vitest", "run"], "test").with_files(),
            ),
        ])
    }

    fn filter_files_for_check(
        &self,
        files: &[String],
        check_type: &str,
        custom_extensions: Option<&[String]>,
    ) -> Vec<String> {
        let filtered = super::runner::base_filter(
            files,
            check_type,
            custom_extensions,
            self.lint_extensions(),
            self.typecheck_extensions(),
            self.test_patterns(),
        );

        if check_type == "test" && filtered.len() == 1 {
            let test_names = staged_typescript_test_names(&self.root, &filtered[0]);
            if test_names.len() == 1 {
                return vec![
                    git_pathspec(&self.root, &filtered[0]),
                    "-t".into(),
                    test_names[0].clone(),
                ];
            }
        }

        filtered
    }

    fn discover_tools(&self, path: &Path, project_config: &ProjectConfig) -> ToolingConfig {
        let mut config = ToolingConfig {
            project_type: self.name().to_string(),
            tools: BTreeMap::new(),
        };
        let package_json = path.join("package.json");
        let Ok(content) = fs::read_to_string(package_json) else {
            return config;
        };
        let Ok(json) = serde_json::from_str::<Value>(&content) else {
            return config;
        };
        let deps = collect_package_deps(&json);
        let scripts = json.get("scripts").and_then(Value::as_object);
        let defaults = self.default_tools();

        if deps_has(&deps, "eslint") || deps_has(&deps, "@eslint/js") {
            let mut tool = tool_from_default(&defaults, "eslint", "lint", project_config);
            tool.pass_files = true;
            config.tools.insert("lint".to_string(), tool);
        } else if deps_has(&deps, "@biomejs/biome") {
            let mut tool = tool_from_default(&defaults, "biome", "lint", project_config);
            tool.pass_files = true;
            config.tools.insert("lint".to_string(), tool);
        }

        if deps_has(&deps, "typescript") {
            let mut tool = tool_from_default(&defaults, "tsc", "typecheck", project_config);
            if scripts.is_some_and(|scripts| scripts.contains_key("typecheck")) {
                tool.command = vec!["npm".into(), "run".into(), "typecheck".into()];
                tool.pass_files = false;
            } else if scripts.is_some_and(|scripts| scripts.contains_key("type-check")) {
                tool.command = vec!["npm".into(), "run".into(), "type-check".into()];
                tool.pass_files = false;
            }
            apply_overrides(&mut tool, project_config);
            config.tools.insert("typecheck".to_string(), tool);
        }

        if deps_has(&deps, "jest") {
            let mut tool = tool_from_default(&defaults, "jest", "test", project_config);
            if scripts.is_some_and(|scripts| scripts.contains_key("test")) {
                tool.command = vec!["npm".into(), "run".into(), "test".into()];
                tool.pass_files = true;
            }
            apply_overrides(&mut tool, project_config);
            config.tools.insert("test".to_string(), tool);
        } else if deps_has(&deps, "vitest") {
            let mut tool = tool_from_default(&defaults, "vitest", "test", project_config);
            if scripts.is_some_and(|scripts| scripts.contains_key("test")) {
                tool.command = vec!["npm".into(), "run".into(), "test".into()];
                tool.pass_files = true;
            }
            apply_overrides(&mut tool, project_config);
            config.tools.insert("test".to_string(), tool);
        }

        config
    }
}

fn staged_typescript_test_names(root: &Path, file: &str) -> Vec<String> {
    let pathspec = git_pathspec(root, file);
    let output = Command::new("git")
        .args(["diff", "--cached", "--unified=0", "--"])
        .arg(pathspec)
        .current_dir(root)
        .output();
    let Ok(output) = output else {
        return Vec::new();
    };
    if !output.status.success() {
        return Vec::new();
    }

    added_typescript_test_names(&String::from_utf8_lossy(&output.stdout))
}

fn git_pathspec(root: &Path, file: &str) -> String {
    let path = Path::new(file);
    if path.is_absolute() {
        return path
            .strip_prefix(root)
            .unwrap_or(path)
            .display()
            .to_string();
    }
    file.to_string()
}

fn added_typescript_test_names(diff: &str) -> Vec<String> {
    diff.lines()
        .filter(|line| line.starts_with('+') && !line.starts_with("+++"))
        .filter_map(|line| typescript_test_name(line.trim_start_matches('+').trim()))
        .collect()
}

fn typescript_test_name(line: &str) -> Option<String> {
    let candidates = ["test", "it"];
    candidates
        .iter()
        .find_map(|prefix| string_arg_after_test_call(line, prefix))
}

fn string_arg_after_test_call(line: &str, prefix: &str) -> Option<String> {
    let mut rest = line.strip_prefix(prefix)?;
    loop {
        rest = rest.trim_start();
        if let Some(after_dot) = rest.strip_prefix('.') {
            let property_len = after_dot
                .chars()
                .take_while(|char| char.is_ascii_alphanumeric() || *char == '_')
                .map(char::len_utf8)
                .sum::<usize>();
            if property_len == 0 {
                return None;
            }
            rest = &after_dot[property_len..];
            continue;
        }
        if let Some(after_paren) = rest.strip_prefix('(') {
            return read_js_string(after_paren.trim_start());
        }
        return None;
    }
}

fn read_js_string(input: &str) -> Option<String> {
    let quote = input.chars().next()?;
    if !matches!(quote, '\'' | '"' | '`') {
        return None;
    }
    let mut escaped = false;
    let mut value = String::new();
    for char in input[quote.len_utf8()..].chars() {
        if escaped {
            value.push(char);
            escaped = false;
            continue;
        }
        if char == '\\' {
            escaped = true;
            continue;
        }
        if char == quote {
            return (!value.is_empty()).then_some(value);
        }
        value.push(char);
    }
    None
}
